mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignDecoder.java

```diff
@@ -24,17 +27,20 @@ public Object decode(Response response, Type type) throws IOException, FeignExce
             throw new MoccaException("Unexpected HTTP response status code: " + response.status());
         }
 
+        final boolean optionalResultType = isParameterizedType(type, Optional.class);
+
         Optional<?> result;
         try (InputStream inputStream = response.body().asInputStream()) {
             if (inputStream == null) {
                 throw new MoccaException("Response does not contain a payload");
             }
 
             final String operationName = MoccaFeignEncoder.getOperationName(response);
-            result = moccaDeserializer.deserialize(inputStream, type, operationName);
+            final Type responseType = optionalResultType ? getInnerType(type) : type;
+            result = moccaDeserializer.deserialize(inputStream, responseType, operationName);
         }
 
-        return result.orElse(null);
+        return optionalResultType ? result : result.orElse(null);
```

Comment:
can we use Optional itself to avoid this conditional check?

Reply:
You mean in the return type of decode method? If that is what you mean, that is not possible, because that is Feign's interface.